### PR TITLE
Add instruction mapping and base RISC-V instructions

### PR DESCRIPTION
**Pull request by xevor11**
Was able to find the common RVI32 and RVI64 instructions between Sail and Qemu, and then created routines to print a json formatted hashmap containing a record for a value for each entry

### CODE REVIEW

File: target/riscv/mapping/instruction_mapping.c [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+}
+
+void to_lowercase(char* str) {
+	for (; *str; ++str) *str = tolower(*str);
+}
+
+char* trim_whitespace(char* str) {
+		char* end;
+		while (isspace((unsigned char)*str)) str++;
+		if (*str == 0)
```
Comment (ThinkOpenly) [inline target/riscv/mapping/instruction_mapping.c:37]: There is a minor inconsistency between the value you use for NULL character here (`0`) and the rest of the code (`'\0'`).

File: target/riscv/mapping/instruction_mapping.c [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+
+		char* lowercase_keyword = strdup(keyword);
+		if (!lowercase_keyword) {
+				fprintf(stderr, "Could not allocate memory for lowercase conversion\n");
+				free(uppercase_keyword);
+				return NULL;
+		}
+		to_lowercase(lowercase_keyword);
+
+		const char* match_start = strstr(sail_function, "match op {");
```
Comment (ThinkOpenly) [inline target/riscv/mapping/instruction_mapping.c:87]: This presumes that the name of the input operand for the instruction enum is "op", which isn't guaranteed. I totally understand that making this generic would be really difficult, though. I presume you'd have to look at the values of the cases within the `match` to ensure they are all instruction enums, then proceed. I'm not convinced this is worthwhile at this stage.

File: target/riscv/mapping/instruction_mapping.c [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+		}
+
+		const char* match_end = strstr(match_start, "};");
+		if (!match_end) {
+				free(uppercase_keyword);
+				free(lowercase_keyword);
+				return NULL;
+		}
+
+		char* keyword_case = strstr(match_start, "RISCV_");
```
Comment (ThinkOpenly) [inline target/riscv/mapping/instruction_mapping.c:101]: This is another presumption, that all instruction enums will begin with "RISCV_", but probably a safe one.

File: target/riscv/mapping/instruction_mapping.c [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+		char* uppercase_keyword = strdup(keyword);
+		if (!uppercase_keyword) {
+				fprintf(stderr, "Could not allocate memory for uppercase conversion\n");
+				return NULL;
+		}
+		to_uppercase(uppercase_keyword);
+
+		char* start = strstr(content, "function clause execute");
+		while (start) {
+				char* end = strstr(start, "RETIRE_SUCCESS");
```
Comment (ThinkOpenly) [inline target/riscv/mapping/instruction_mapping.c:167]: I wonder if there are cases where the function contains conditional code which contains "RETIRE_SUCCESS"?
```
if something
  RETIRE_SUCCESS
do-something-else
RETIRE_SUCCESS
```

**Review by ThinkOpenly — COMMENTED** [review]
General comments, mostly subjective, and a few questions:
- You are parsing the Sail code, which is fine, but I wonder if it would be easier to parse the JSON? In that way, you can quickly get to the function body of the instruction. You'd also only have to import a single file from "upstream": the JSON file for all of the Sail files.
- I really like how you have managed to eliminate the code which pertains to other mnemonics, so the resulting function body is really just the code for the specific mnemonic.
- It will be most comfortable for reviewers if the code style matched existing code in the upstream repository, so 4-character indentation for blocks, and continuation lines align with a logical break in the first or prior line.

**Comment by xevor11** [comment]
Done

**Pull request merged**